### PR TITLE
OSL Plugin fixes

### PR DESCRIPTION
**Pull request by usakhelo**
Hi,
Fixes in this PR:

1. Shader files in separate directories
2. Shaders from env. variable path override shaders in plugin directory
3. Texture buttons in OSL material plugins are update properly
4. Parameters are hidden based on special metadata token
5. Vector parameter connection works
6. Viewport shading for OSL materials

Thanks for review.

### CODE REVIEW

File: src/appleseed-max-impl/appleseedoslplugin/oslmaterial.cpp [hunk 1]
```diff
@@ -47,6 +48,7 @@
 
 // 3ds Max headers.
 #include <iparamm2.h>
+#include <hsv.h>
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslmaterial.cpp]: Order alphabetically.

File: src/appleseed-max-impl/appleseedoslplugin/oslmaterial.cpp [hunk 2]
```diff
@@ -224,17 +254,25 @@ void OSLMaterial::SetSubTexmap(int i, Texmap* tex_map)
...
-        m_pblock->SetValue(param_id, 0, tex_map);
-
-        if (m_pblock->GetParameterType(param_id - 1) == TYPE_POINT3)
+        const int param_id = m_texture_id_map[i].first;
+        if (m_has_bump_params && i == m_texture_id_map.size() - 1)
         {
-            IParamMap2* map = m_pblock->GetMap(0);
-            if (map != nullptr)
-                map->SetText(param_id, tex_map == nullptr ? L"" : L"M");
+            BOOL result = m_bump_pblock->SetValue(param_id, 0, tex_map);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslmaterial.cpp]: Remove `result` variable (unused).

File: src/appleseed-max-impl/appleseedoslplugin/oslmaterial.cpp [hunk 3]
```diff
@@ -224,17 +254,25 @@ void OSLMaterial::SetSubTexmap(int i, Texmap* tex_map)
...
+            m_class_desc->GetParamBlockDesc(1)->InvalidateUI(param_id);
         }
+        else
+        {
+            BOOL result = m_pblock->SetValue(param_id, 0, tex_map);
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslmaterial.cpp]: Remove `result`.

File: src/appleseed-max-impl/appleseedoslplugin/oslmaterial.cpp [hunk 4]
```diff
@@ -347,27 +389,33 @@ Color OSLMaterial::GetAmbient(int mtlNum, BOOL backFace)
 
 Color OSLMaterial::GetDiffuse(int mtlNum, BOOL backFace)
 {
-    return Color(0.0f, 0.0f, 0.0f);
+    Mtl* mtl = nullptr;
+    Interval iv;
+    if (m_submaterial_map.size() > 0)
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslmaterial.cpp]: if (!m_submaterial_map.empty())

is better.

File: src/appleseed-max-impl/appleseedoslplugin/oslmaterial.cpp [hunk 5]
```diff
@@ -347,27 +389,33 @@ Color OSLMaterial::GetAmbient(int mtlNum, BOOL backFace)
...
 float OSLMaterial::GetShinStr(int mtlNum, BOOL backFace)
 {
-    return 0.0f;
+    return get_viewport_shininess_amount();
 }
 
 float OSLMaterial::GetXParency(int mtlNum, BOOL backFace)
 {
-    return 0.0f;
+    float transparency = get_viewport_transparency_amount();
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslmaterial.cpp]: Make const.

File: src/appleseed-max-impl/appleseedoslplugin/oslparamdlg.cpp [hunk 6]
```diff
@@ -132,59 +132,76 @@ void OSLParamDlg::add_ui_parameter(
...
-
-      case MaxParam::String:
-        dialog_template.AddComponent((LPCSTR)"CustEdit", (LPCSTR)"Edit", WS_VISIBLE | WS_TABSTOP, NULL, Col2X, y_pos, TexButtonWidth, EditHeight, ctrl_id++);
-        break;
-
-      case MaxParam::Closure:
-        dialog_template.AddComponent((LPCSTR)"CustButton", (LPCSTR)"", WS_VISIBLE, NULL, Col3X, y_pos, TexButtonWidth, EditHeight, ctrl_id++);
+        switch (max_param.m_param_type)
+        {
+        case MaxParam::Float:
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslparamdlg.cpp]: Half-indent `case` statements.

File: src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.cpp [hunk 7]
```diff
@@ -102,6 +102,7 @@ namespace
 OSLParamInfo::OSLParamInfo(const asf::Dictionary& paramInfo)
   : hasDefault(false)
   , divider(false)
+  , lockGeom(true)
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshadermetadata.cpp]: I think we should change `OSLParamInfo` to use the same naming conventions as the rest of the project (`m_` prefix and snake_case for member variables).

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp [hunk 8]
```diff
@@ -74,28 +103,28 @@ namespace
...
-            if (tex_map != nullptr)
-                map->SetText(id, is_short ? L"M" : tex_map->GetFullName());
-            else
-                map->SetText(id, is_short ? L"" : L"None");
+            set_button_text(pblock, id);
         }
     };
     
+    class BumpTextureAccessor
+        : public PBAccessor
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: Only half-indentation here.

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp [hunk 9]
```diff
@@ -74,28 +103,28 @@ namespace
...
+    {
+    public:
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.cpp]: Half-indent.

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.h [hunk 10]
```diff
@@ -46,18 +47,26 @@ class OSLShaderRegistry
...
     int get_size() const;
     void create_class_descriptors();
-    void add_parameter(
+    void add_const_parameter(
         ParamBlockDesc2*        pb_desc,
         const OSLParamInfo&     osl_param,
         MaxParam&               param_info,
         IdNameMap&              string_map,
+        const int               param_id,
+        int&                    ctrl_id,
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.h]: I find it a bit surprising that `add_const_parameter()` needs to modify `ctrl_id` and `string_id` but `add_input_parameter()` don't. Is there a fundamental reason why?

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.h [hunk 11]
```diff
@@ -37,6 +37,7 @@
 
 // Standard headers.
 #include <vector>
+#include <string>
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.h]: Order alphabetically.

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.h (hunk 10 above)
Comment (usakhelo) [reply src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.h]: Yes.. I've tried to make `add_parameter()` prettier by splitting in two functions and moving some increments outside.
Thing is that `add_const_parameter()` creates variable number of controls (like three point3 spinners) and string objects (for comboboxes).
`add_input_parameter()` add fixed number of controls, that's why it doesn't change ctrl_id from inside the function body.

**Comment by usakhelo** [comment]
@dictoon I've fixed the comments and add one more commit where broken output parameter is fixed and string inputs connected properly. That makes as_texture shader work.

File: src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.h (hunk 10 above)
Comment (dictoon) [reply src/appleseed-max-impl/appleseedoslplugin/oslshaderregistry.h]: Ok, thanks for the explanation.

**Pull request merged**